rs/nns/governance/src/voting.rs

```diff
@@ -161,47 +359,79 @@ impl ProposalVotingStateMachine {
         &mut self,
         neuron_store: &mut NeuronStore,
         ballots: &mut HashMap<u64, Ballot>,
+        is_over_instructions_limit: fn() -> bool,
     ) {
-        while let Some(neuron_id) = self.neurons_to_check_followers.pop_first() {
-            self.add_followers_to_check(neuron_store, neuron_id, self.topic);
-        }
-
-        // Memory optimization, will not cause tests to fail if removed
-        retain_neurons_with_castable_ballots(&mut self.followers_to_check, ballots);
-
-        while let Some(follower) = self.followers_to_check.pop_first() {
-            let vote = match neuron_store.neuron_would_follow_ballots(follower, self.topic, ballots)
-            {
-                Ok(vote) => vote,
-                Err(e) => {
-                    // This is a bad inconsistency, but there is
-                    // nothing that can be done about it at this
-                    // place.  We somehow have followers recorded that don't exist.
-                    eprintln!("error in cast_vote_and_cascade_follow when gathering induction votes: {:?}", e);
-                    Vote::Unspecified
+        let mut actions_performed = 0;
+        let check_after_number_actions = 50;
```

Comment:
It's still unclear (to me) why this is needed, since IIUC checking instructions against a limit should cost 100-1000 instructions and therefore much fewer than any of the operations here. It would be great to avoid those unnecessary complexity if possible.

Reply:
Hm - here are the performance numbers when checking every time for tests that are affected:
```
---------------------------------------------------

Benchmark: cascading_vote_stable_neurons_with_heap_index
  total:
    instructions: 347.35 M (regressed by 171.95%)
    heap_increase: 0 pages (no change)
    stable_memory_increase: 128 pages (no change)

---------------------------------------------------

Benchmark: cascading_vote_stable_everything
  total:
    instructions: 369.54 M (regressed by 146.48%)
    heap_increase: 0 pages (no change)
    stable_memory_increase: 128 pages (no change)

---------------------------------------------------

Benchmark: centralized_following_all_stable
  total:
    instructions: 171.33 M (regressed by 415.17%)
    heap_increase: 0 pages (no change)
    stable_memory_increase: 128 pages (no change)

---------------------------------------------------

Doesn't make a lot of sense, actually, as you'd expect ALL the voting to perform worse, even when using heap, as they all would be checking just as much....

Reply:
One clue - instruction_counter() inside canbench doesn't allow you to get an answer relative to the start of all the canbench activity, which means that our checks will be causing it to constantly store and retrieve the state machine in stable storage.

What's weird, however, is that this isn't happening in some of the other cases hitting the same code paths... perhaps they somehow fall under the limit overall, so that the loading/saving of the ProposalVotingStateMachine is avoided.